File: web/src/date.ts
// The board's days live in ONE time zone for every user (the server's
// AEMAN_TZ, delivered via /api/config): without it a user east of the server
// crosses their local midnight and starts seeing deferred "tomorrow" cards on
// today's board. Unset = the browser's own zone (single-zone teams).
let boardTz: string | undefined;

/** setBoardTimezone installs the board's day zone (an IANA name, "" = local). */
export function setBoardTimezone(tz: string | undefined): void {
  boardTz = tz && tz !== "Local" ? tz : undefined;
}

// en-CA formats as yyyy-mm-dd; an invalid zone falls back to the browser's.
function dayInBoardZone(d: Date): string {
  try {
    return new Intl.DateTimeFormat("en-CA", {
      timeZone: boardTz,
      year: "numeric",
      month: "2-digit",
      day: "2-digit",
    }).format(d);
  } catch {
    return new Intl.DateTimeFormat("en-CA", {
      year: "numeric",
      month: "2-digit",
      day: "2-digit",
    }).format(d);
  }
}

/** todayIso returns today's date as a yyyy-mm-dd string in the BOARD zone. */
export function todayIso(): string {
  return dayInBoardZone(new Date());
}

/** localDateIso returns the board-zone yyyy-mm-dd date for an ISO timestamp. */
export function localDateIso(iso: string): string {
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) {
    return "";
  }
  return dayInBoardZone(d);
}

/** activeOnDay is true when `day` falls within a card's visible span. The span
 * runs from `start` (the card's day) to the later of start/finish, so a card
 * whose start day is past its sprint (added on a later day) still shows on its
 * day, while a carried card shows from its origin through the current sprint. A
 * card with neither date is never on the day board. */
export function activeOnDay(
  start: string | undefined,
  finish: string | undefined,
  day: string,
): boolean {
  const from = start || finish;
  if (!from) {
    return false;
  }
  const to = finish && finish > from ? finish : from;
  return from <= day && day <= to;
}

/** daysSince returns whole days between an ISO timestamp's local date and the
 * `asOf` day (a yyyy-mm-dd string, default today). */
export function daysSince(iso: string | undefined, asOf?: string): number {
  if (!iso) {
    return 0;
  }
  const then = localDateIso(iso);
  if (!then) {
    return 0;
  }
  const [py, pm, pd] = then.split("-").map(Number);
  const [ty, tm, td] = (asOf || todayIso()).split("-").map(Number);
  const ms = Date.UTC(ty, tm - 1, td) - Date.UTC(py, pm - 1, pd);
  return Math.max(0, Math.round(ms / 86_400_000));
}

/** mondayOf returns the yyyy-mm-dd Monday of the week containing `iso`. */
export function mondayOf(iso: string): string {
  const [y, m, d] = iso.split("-").map(Number);
  if (!y || !m || !d) {
    return iso;
  }
  const dt = new Date(y, m - 1, d);
  dt.setDate(dt.getDate() - ((dt.getDay() + 6) % 7));
  const mm = String(dt.getMonth() + 1).padStart(2, "0");
  const dd = String(dt.getDate()).padStart(2, "0");
  return `${dt.getFullYear()}-${mm}-${dd}`;
}

/** addDays shifts a yyyy-mm-dd date by delta days, returning yyyy-mm-dd. */
export function addDays(iso: string, delta: number): string {
  const [y, m, d] = iso.split("-").map(Number);
  if (!y || !m || !d) {
    return iso;
  }
  const dt = new Date(y, m - 1, d + delta);
  const mm = String(dt.getMonth() + 1).padStart(2, "0");
  const dd = String(dt.getDate()).padStart(2, "0");
  return `${dt.getFullYear()}-${mm}-${dd}`;
}

/** weeksBetween counts whole weeks from Monday a to Monday b (0 = same week).
 *  Both boards measure a slot's span with it: the Project board to place a
 *  row, the Team board to keep that span when the plan panel moves a slot. */
export function weeksBetween(a: string, b: string): number {
  const [ay, am, ad] = a.split("-").map(Number);
  const [by, bm, bd] = b.split("-").map(Number);
  return Math.round(
    (Date.UTC(by, bm - 1, bd) - Date.UTC(ay, am - 1, ad)) / (7 * 86400000),
  );
}

File: pkg/board/date.go
package board

import "time"

// isoLayout is the yyyy-mm-dd date layout the boards orient by.
const isoLayout = "2006-01-02"

// TodayIso returns today's date as a local yyyy-mm-dd string. It mirrors todayIso
// in web/src/date.ts.
// boardLocation is the time zone the BOARD's days live in — one zone for
// every user and the server, or a Moscow teammate crossing their local
// midnight starts seeing "tomorrow's" cards while the lead is still on
// today's board. Defaults to the server's local zone; SetLocation installs
// the configured one (AEMAN_TZ).
var boardLocation = time.Local

// SetLocation installs the board time zone (name per IANA, e.g.
// "Europe/Berlin"). An unknown name is reported and leaves the zone as-is.
func SetLocation(name string) error {
	loc, err := time.LoadLocation(name)
	if err != nil {
		return err
	}
	boardLocation = loc
	return nil
}

// LocationName reports the installed board time zone (IANA name or "Local").
func LocationName() string {
	return boardLocation.String()
}

func TodayIso() string {
	return time.Now().In(boardLocation).Format(isoLayout)
}

// LocalDateIso returns the local yyyy-mm-dd date for an ISO timestamp, or "" when
// the timestamp cannot be parsed. It mirrors localDateIso in web/src/date.ts.
func LocalDateIso(iso string) string {
	t, ok := parseTimestamp(iso)
	if !ok {
		return ""
	}
	return t.In(boardLocation).Format(isoLayout)
}

// ActiveOnDay reports whether `day` falls within a card's [start, finish] range.
// A missing bound collapses to the other date, so a card with a single date is
// active only on that day and a card with neither date is never on the day board.
// It mirrors activeOnDay in web/src/date.ts (from = start||finish,
// to = finish||start, from <= day <= to).
func ActiveOnDay(start, finish, day string) bool {
	from := start
	if from == "" {
		from = finish
	}
	if from == "" {
		return false
	}
	// The span runs from `start` to the later of start/finish, so a card whose
	// start day is past its sprint (added on a later day) still shows on its day.
	to := from
	if finish > from {
		to = finish
	}
	return from <= day && day <= to
}

// DaysSince returns the whole days between an ISO timestamp's local date and the
// `asOf` day (a yyyy-mm-dd string; "" means today). It never goes negative. It
// mirrors daysSince in web/src/date.ts.
func DaysSince(iso, asOf string) int {
	if iso == "" {
		return 0
	}
	then := LocalDateIso(iso)
	if then == "" {
		return 0
	}
	if asOf == "" {
		asOf = TodayIso()
	}
	t1, err1 := time.Parse(isoLayout, then)
	t2, err2 := time.Parse(isoLayout, asOf)
	if err1 != nil || err2 != nil {
		return 0
	}
	days := int(t2.Sub(t1) / (24 * time.Hour))
	if days < 0 {
		return 0
	}
	return days
}

// MondayOf returns the yyyy-mm-dd Monday of the week containing `iso`, or `iso`
// unchanged when it is not a parseable date. It mirrors mondayOf in
// web/src/date.ts.
func MondayOf(iso string) string {
	t, err := time.Parse(isoLayout, iso)
	if err != nil {
		return iso
	}
	// time.Weekday and JS getDay both run Sunday=0..Saturday=6, so (wd+6)%7 is the
	// number of days back to Monday.
	offset := (int(t.Weekday()) + 6) % 7
	return t.AddDate(0, 0, -offset).Format(isoLayout)
}

// AddDays shifts a yyyy-mm-dd date by delta days, returning yyyy-mm-dd, or `iso`
// unchanged when it is not a parseable date. It mirrors addDays in
// web/src/date.ts.
func AddDays(iso string, delta int) string {
	t, err := time.Parse(isoLayout, iso)
	if err != nil {
		return iso
	}
	return t.AddDate(0, 0, delta).Format(isoLayout)
}

// parseTimestamp parses an ISO timestamp the way the frontend's `new Date(iso)`
// does for the inputs aeman sees: an RFC 3339 timestamp or a bare yyyy-mm-dd.
func parseTimestamp(iso string) (time.Time, bool) {
	if iso == "" {
		return time.Time{}, false
	}
	for _, layout := range []string{time.RFC3339Nano, time.RFC3339, "2006-01-02T15:04:05", isoLayout} {
		if t, err := time.Parse(layout, iso); err == nil {
			return t, true
		}
	}
	return time.Time{}, false
}
